src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -355,37 +355,51 @@ internal static void UnshimResponseStream(ResponseCachingContext context)
         internal static bool ContentIsNotModified(ResponseCachingContext context)
         {
             var cachedResponseHeaders = context.CachedResponseHeaders;
-            var ifNoneMatchHeader = context.TypedRequestHeaders.IfNoneMatch;
+            var ifNoneMatchHeader = context.HttpContext.Request.Headers[HeaderNames.IfNoneMatch];
 
-            if (ifNoneMatchHeader != null)
+            if (!StringValues.IsNullOrEmpty(ifNoneMatchHeader))
             {
-                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any))
+                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any.Tag))
                 {
                     context.Logger.LogNotModifiedIfNoneMatchStar();
                     return true;
                 }
 
-                if (cachedResponseHeaders.ETag != null)
+                if (!StringValues.IsNullOrEmpty(cachedResponseHeaders[HeaderNames.ETag]))
                 {
-                    foreach (var tag in ifNoneMatchHeader)
+                    EntityTagHeaderValue eTag;
+                    if (EntityTagHeaderValue.TryParse(cachedResponseHeaders[HeaderNames.ETag], out eTag))
                     {
-                        if (cachedResponseHeaders.ETag.Compare(tag, useStrongComparison: false))
+                        foreach (var tag in ifNoneMatchHeader)
                         {
-                            context.Logger.LogNotModifiedIfNoneMatchMatched(tag);
-                            return true;
+                            EntityTagHeaderValue requestETag;
+                            if (EntityTagHeaderValue.TryParse(tag, out requestETag) &&
+                                eTag.Compare(requestETag, useStrongComparison: false))
+                            {
+                                context.Logger.LogNotModifiedIfNoneMatchMatched(requestETag);
+                                return true;
+                            }
                         }
                     }
                 }
             }
             else
             {
-                var ifUnmodifiedSince = context.TypedRequestHeaders.IfUnmodifiedSince;
-                if (ifUnmodifiedSince != null)
+                var ifUnmodifiedSince = context.HttpContext.Request.Headers[HeaderNames.IfUnmodifiedSince];
+                if (!StringValues.IsNullOrEmpty(ifUnmodifiedSince))
                 {
-                    var lastModified = cachedResponseHeaders.LastModified ?? cachedResponseHeaders.Date;
-                    if (lastModified <= ifUnmodifiedSince)
+                    DateTimeOffset modified;
+                    if (!HttpHeaderParsingHelpers.TryParseHeaderDate(cachedResponseHeaders[HeaderNames.LastModified], out modified) &&
```

Comment:
I have this fixed in my clean up commit but just to note, both sides of the `&&` operator are evaluated regardless of whether they are true or false. See example here: https://gist.github.com/JunTaoLuo/87f8b2e727c956333fa036f01b3275af